erigon-lib/commitment/metrics.go

```diff
@@ -0,0 +1,421 @@
+			metrics.TotalFoldingTimeInc(startUnfold, plainKey)
+		}
+	}
+	return func() {}
+}
+
+func (metrics *Metrics) TotalUnfoldingTimeInc(t time.Time, plainKey []byte) {
+	if collectCommitmentMetrics {
+		metrics.totalUnfoldingTime += time.Since(t)
+		metrics.Accounts.TotalUnfoldingTimeInc(plainKey, t)
+	}
+}
+
+func (metrics *Metrics) TotalFoldingTimeInc(t time.Time, plainKey []byte) {
+	if collectCommitmentMetrics {
+		metrics.totalFoldingTime += time.Since(t)
+		metrics.Accounts.TotalFoldingTimeInc(plainKey, t)
+	}
+}
+
+func (metrics *Metrics) TotalProcessingTimeInc(t time.Time) {
+	if collectCommitmentMetrics {
+		metrics.totalProcessingTime += time.Since(t)
+	}
+}
+
+func NewAccounts() *AccountMetrics {
+	return &AccountMetrics{
+		AccountStats: make(map[string]*AccountStats),
+	}
+}
+
+type AccountStats struct {
+	AccountUpdates     uint64
+	StorageUpates      uint64
+	LoadBranch         uint64
+	LoadAccount        uint64
+	LoadStorage        uint64
+	Unfolds            uint64
+	TotalUnfoldingTime time.Duration
+	Folds              uint64
+	TotalFoldingTime   time.Duration
+}
+
+type AccountMetrics struct {
+	m sync.Mutex
+	// will be separate value for each key in parallel processing
+	AccountStats map[string]*AccountStats
+}
+
+func (processAccount *AccountMetrics) Headers() []string {
+	return []string{
+		"account",
+		"storage updates",
+		"loading branch",
+		"loading account",
+		"loading storage",
+		"total unfolds",
+		"total unfolding time (μs)",
+		"total folds",
+		"total folding time (μs)",
+	}
+}
+
+func (processAccount *AccountMetrics) Values() [][]string {
+	processAccount.m.Lock()
+	defer processAccount.m.Unlock()
+	// + 1 to add one empty line between "process" calls
+	values := make([][]string, len(processAccount.AccountStats)+1, len(processAccount.AccountStats)+1)
+	ind := 1
+	for i, account := range processAccount.AccountStats {
+		values[ind] = []string{
+			i,
+			strconv.FormatUint(account.StorageUpates, 10),
+			strconv.FormatUint(account.LoadBranch, 10),
+			strconv.FormatUint(account.LoadAccount, 10),
+			strconv.FormatUint(account.LoadStorage, 10),
+			strconv.FormatUint(account.Unfolds, 10),
+			strconv.Itoa(int(account.TotalUnfoldingTime.Microseconds())),
+			strconv.FormatUint(account.Folds, 10),
+			strconv.Itoa(int(account.TotalFoldingTime.Microseconds())),
+		}
+		ind++
+	}
+	return values
+}
+
+func (processAccount *AccountMetrics) Reset() {
+	processAccount.m.Lock()
+	defer processAccount.m.Unlock()
+	processAccount.AccountStats = make(map[string]*AccountStats)
+}
+
+func (processAccount *AccountMetrics) Updates(plainKey []byte) {
+	if !collectCommitmentMetrics {
+		return
+	}
+	processAccount.UpdatesInc(plainKey)
+	processAccount.UpdatesStorageInc(plainKey)
+}
+
+func (processAccount *AccountMetrics) UpdatesInc(plainKey []byte) {
+	if collectCommitmentMetrics && plainKey != nil {
+		processAccount.m.Lock()
+		defer processAccount.m.Unlock()
+		account := hex.EncodeToString(plainKey[0:20])
```

Comment:
FYI: can do string(pk) or even better - `toStringZeroCopy` if it's within this package. 

There is NO chance that during one run of `Processing()` you will met same account twice
so counter will be always 1 or addr not exists. 
Therefore UpdatesInc is pointless at all as well as AccountUpdates field there
Therefoe need to move 
```
if !collectCommitmentMetrics {
		return
}
```
into UpdatesStorageInc and remove `.Updates` and `.UpdatesInc` methods.

You do
```
	if len(plainKey) == hph.accountKeyLen {
		hph.metrics.addressKeys.Add(1)
```
which also triggered only when Update is processed. Therefore does not make sense to have `.UpdateStorageInc` method and this line separated.

Reply:
also pls not ignore the fact that prefixes could be of different length so :20 could just not work and panic in tests.